### ISSUE TITLE
Hardcoded model classes in constructor

### ISSUE THREAD

**Issue by waaghals** [Sylius/Sylius#2371]
**Sylius/src/Sylius/Component/Core/Model/Product.php L:68-76**
```php
    /**
     * Constructor.
     */
    public function __construct()
    {
        parent::__construct();
        $this->setMasterVariant(new ProductVariant());
        $this->taxons = new ArrayCollection();
        $this->variantSelectionMethod = self::VARIANT_SELECTION_CHOICE;
    }
```

This creates problems when I override the `ProductVariant` model. The Form expects `My\Namespace\ProcuctVariant` but gets `Core\Model\ProductVariant`.

Either the MasterVariant should be empty or the variant should be created using the repository.

**Comment by stloyd** [comment]
We know this, it's really old issue, yet it can't be fixed properly without introducing proper resource managers (we have 2 PRs being reviewed as they state in two directions how to solve the problem).

**Comment by axelvnk** [comment]
+1 for this issue. I found this out this morning when trying to extend the ProductVariant class. One option would be to have a parameter in the constructor like so
```
 __construct(ProductVariantInterface $productVariant)
{
    this->setMasterVariant($productVariant);
    ...
}
````

But this will of course be more "difficult" with basic usage of the ProductVariant model.

**Comment by pjedrzejewski** [comment]
Reported in #480 and will be fixed just like #3480.
